File: 84/solution_stack.rb
# monotonic stack

# @param {Integer[]} heights
# @return {Integer}
def largest_rectangle_area(heights)
    heights.push 0
    n = heights.size
    s = []
    result = i = 0
    
    while i < n
        if s.empty? || heights[i] >= heights[s[-1]]
            s.push(i)
            i += 1
        else
            h = heights[s[-1]]; s.pop
            w = s.empty? ? i : i - s[-1] - 1
            result = [result, h * w].max
        end
    end
    
    result
end

File: 503/solution_stack.cpp
// monotonic stack

class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans(n);
        stack<int> s;
        
        for(int i = 2 * n - 1; i >= 0; i--) {
            while(!s.empty() && s.top() <= nums[i % n])
                s.pop();
            ans[i % n] = s.empty() ? -1 : s.top();
            s.push(nums[i % n]);
        }
        
        return ans;
    }
};
